### DISCUSSION TITLE
IDs on subdocuments

### DISCUSSION THREAD

**Discussion by luzzardik** [typegoose/typegoose#917] [Q&A] [answered]
Hi,

I'm trying to use Subdocuments, but would there be an _id generated automatically by default ?
What should I do to disable them ?

**Comment by smolinari** [comment]
@luzzardik 

https://typegoose.github.io/typegoose/docs/api/decorators/prop/#_id

But also read this:

https://typegoose.github.io/typegoose/docs/guides/advanced/reference-other-classes/#reference-other-classes-with-different-_id-type

and this:

https://typegoose.github.io/typegoose/docs/guides/advanced/change-id-type

Scott

**Reply by luzzardik** [reply]
Hi, thanks ! How would it work on Arrays of Subdocuments ? I can't find it in the docs.

**Reply by smolinari** [reply]
How would what work?

Scott

**Reply by luzzardik** [reply]
`
@prop({ type: Subdoc, default: [], _id: false })
subdoc_array!: Types.Array<Subdoc>;
`
Here, would the _id: false propagate to all subdocs in the array ?

**Reply by hasezoey** [reply]
>  Here, would the _id: false propagate to all subdocs in the array ?

it should, but `_id: false` can also be set in [schema-options](https://mongoosejs.com/docs/api/schema.html) (which is set via [`@modelOptions({ schemaOptions: {} })`](https://typegoose.github.io/typegoose/docs/api/decorators/model-options#schemaoptions)

**Reply by luzzardik** [reply]
Oh OK ! That wasn't clear to be honest as @modelOptions was for me only used when calling getModelForClass, is it not the case ?

**Reply by smolinari** [reply]
All Typegoose decoration is only effective at the point `getModelForClass` is called AFAIK. 

Scott

**Reply by hasezoey** [reply, marked as answer]
> was for me only used when calling getModelForClass, is it not the case ?
> All Typegoose decoration is only effective at the point getModelForClass is called AFAIK.

to clarify, it is effective if you call `buildSchema`, not just `getModelForClass`; ie the place where the schemas actually get build.

or if i somehow mis-understood you question; i want to note that `@modelOptions` is a *class* decorator, not a *property* decorator.

also maybe important to know is that when you use `@modelOptions` to set `_id: false`, then anywhere you use that class (schema) you dont need to set `_id: false` in `@prop`, but also note that those options *do not* affect `_id` if the class (schema) is used as the document root (or said in different words, `_id: false` is only effective if used as / on a subdocument)

**Reply by smolinari** [reply]
> to clarify, it is effective if you call buildSchema, not just getModelForClass; ie the place where the schemas actually get build.

Ah yes. I don't ever use `buildSchema`, so I forgot about it. :grin: 

Scott
